examples/textbooks/generate_fineweb.py

```diff
@@ -0,0 +1,249 @@
+                        stop_sequences=STOP_SEQ,
+                        temperature=args.temperature,
+                        top_p=args.top_p,
+                        top_k=args.top_k,
+                        repetition_penalty=args.repetition_penalty,
+                    )
+                    for stop_seq in STOP_SEQ:
+                        if completion.endswith(stop_seq):
+                            completion = completion[: -len(stop_seq)].rstrip()
+                    token_length += len(tokenizer.encode(completion))
+                    sample["completion"] = completion
+                    sample["token_length"] = token_length
+                    return sample
+
+            except Exception as e:
+                attempt += 1
+                if attempt < MAX_RETRIES:
+                    print(
+                        f"Request failed, retrying in {RETRY_DELAY} seconds... (Attempt {attempt}/{MAX_RETRIES})"
+                    )
+                    await asyncio.sleep(RETRY_DELAY)
+                else:
+                    print(
+                        f"Max retries reached. Failed to process the request with error {str(e)}."
+                    )
+                    sample["completion"] = ""
+                    sample["token_length"] = 0
+                    return sample
+
+    async def main():
+        start_time = time.time()
+        total_tokens = 0
+        saving_time = 0
+
+        wandb.init(
+            project="synthetic_data",
+            entity=args.wandb_username,
+            name=repo_id.split("/")[1],
+        )
+        wandb.config.update(asdict(args))
+
+        repo_id = (
+            f"{args.repo_id}_{args.prompt_column}"
+            if args.prompt_column not in args.repo_id
+            else args.repo_id
+        )
+        checkpoint_dir = f"{args.checkpoint_path}/{repo_id.split('/')[1]}/data"
+        os.makedirs(checkpoint_dir, exist_ok=True)
+        print(f"Will be saving at {checkpoint_dir}")
+
+        total_samples = len(ds)
+        for i in range(0, total_samples, args.checkpoint_interval):
+            batch_time = time.time()
+            # Processing a chunk
+            print(
+                f"Processing chunk {int(i/args.checkpoint_interval)}/{int(total_samples/args.checkpoint_interval)}"
+            )
+            end_index = min(i + args.checkpoint_interval, total_samples)
+            chunk = ds.select(range(i, end_index))
+            chunk_results = await tqdm_asyncio.gather(
+                *(process_text(sample) for sample in chunk)
+            )
+            # Save the chunk results and log throughput
+            temp_time = time.time()
+            time_per_chunk = temp_time - batch_time
+            checkpoint_path = os.path.join(checkpoint_dir, f"checkpoint_{i}.json")
+            with open(checkpoint_path, "w") as f:
+                json.dump(chunk_results, f)
+
+            df = pd.DataFrame.from_records(chunk_results)
```

Comment:
This is totally fine — something I found useful in #24 was

https://github.com/huggingface/llm-swarm/blob/fff37e6a21aedd8ff1f0ac970bfc4086e10be433/examples/openhermes-preference/generate.py#L144-L145

Maybe `Dataset` would be better? As it should handle partitioning / sharding automatically vs json.

Reply:
Oh I didn't know the `from_list` existed, thanks! I changed it, but I kept `to_json` instead of `save_to_disk` since this is already sharded + we can easily load the dataset later to not keep all the generations in memory